[FLINK-33677][core] Remove flink-conf.yaml from flink dist.

JunRuiLee opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

 This pr introduces breaking changes that will be included in FLINK-2.0: Remove flink-conf.yaml from flink dist.


## Brief change log

Remove flink-conf.yaml from flink dist.


## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhuzhurk commented on flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java:
```diff
@@ -45,6 +45,11 @@ public final class GlobalConfiguration {
+     * @deprecated This file is deprecated and is only used to help users migrate their legacy
+     *     configuration files to the new configuration file `config.yaml`.
+     */
+    @Deprecated @VisibleForTesting
```
How about to introduce a separate `ConfigurationFileMigrationUtils` class? It can be the only place to host logics which recognizes the legacy file.

zhuzhurk commented on flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java:
```diff
@@ -143,31 +155,54 @@ public static Configuration loadConfiguration(
-        if (!yamlConfigFile.exists()) {
-            yamlConfigFile = new File(confDirFile, FLINK_CONF_FILENAME);
+        if (loadLegacyConfigFile) {
+            File yamlConfigFile = new File(confDirFile, LEGACY_FLINK_CONF_FILENAME);
```
I prefer that the `GlobalConfiguration` code no longer see nor operate on the legacy config file, just like mentioned in the comment above.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ConfigurationParserUtils.java:
```diff
@@ -230,13 +228,9 @@ public static List<String> migrateLegacyConfigurationToStandardYaml(
             throw e;
         }
 
-        checkState(
```
Why removing this check?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java:
```diff
@@ -298,8 +298,15 @@ private void testWriteConfigurationAndReloadInternal(boolean standardYaml) throw
+            loadedFlinkConfig =
+                    GlobalConfiguration.loadConfiguration(flinkConfDir.getAbsolutePath());
+        } else {
+            loadedFlinkConfig =
```
We should not test cases that will not happen in production.

JunRuiLee commented:
@zhuzhurk, thanks for the review, I've updated this pr accordingly, PTAL.

JunRuiLee commented:
I've updated the PR to completely remove the standard YAML flag from the code. 
Additionally, I've removed the setStandardYaml calls from `RocksDBStateBackendConfigTest#testConfigureForStCompressionPerLevel `and `ForStStateBackendConfigTest#testConfigureForStCompressionPerLevel`, as discussed with @Zakelly. Although this case is inconsistent with previous behavior, it remains meaningful.

JunRuiLee commented:
@flinkbot run azure

Zakelly commented:
> I've updated the PR to completely remove the standard YAML flag from the code. Additionally, I've removed the setStandardYaml calls from `RocksDBStateBackendConfigTest#testConfigureForStCompressionPerLevel `and `ForStStateBackendConfigTest#testConfigureForStCompressionPerLevel`, as discussed with @Zakelly. Although this case is inconsistent with previous behavior, it remains meaningful.

+1 for this part.

zhuzhurk commented on flink-python/pyflink/pyflink_gateway_server.py:
```diff
@@ -46,32 +46,13 @@ def on_windows():
+            flat_config = flatten_config(config)
+            return flat_config.get(key, default_value)
 
     # If neither file exists, return the default value
```
The comment is outdated.

zhuzhurk commented on flink-core/src/test/java/org/apache/flink/configuration/ConfigurationFileMigrationUtilsTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    // We allow malformed YAML files if loaded legacy flink conf
```
loaded -> loading

zhuzhurk reviewed (commented):
Looks good to me except for 2 minor comments.
Would you fix it, squash the commits and rebase it onto the latest master?

JunRuiLee commented:
Thanks @zhuzhurk , I've updated this pr accordingly, PTAL.

zhuzhurk reviewed (approved):
LGTM.
